### DISCUSSION TITLE
🔍 SSG/SSR software feature comparison

### DISCUSSION THREAD

**Discussion by mohsenkhanpour** [alephjs/aleph.js#426] [General]
| Framework | [Aleph.js ](https://alephjs.org/) | [Next.js](https://nextjs.org/)/[Nuxt.js ](https://nuxtjs.org/) | [Gatsby.js](https://www.gatsbyjs.com/)/[Gridsome ](https://gridsome.org/) | [Sapper](https://sapper.svelte.dev/) | [eleventy](https://www.11ty.dev/  ) |
|-|-|-|-|-|-|
| Templating | Alef/React/Preact/Markdown | React/Vue | React/Preact/Vue | Svelte | HTML/Markdown/JavaScript/Liquid/Nunjucks/Handlebars/Mustache/EJS/Haml/Pug/[Vue](https://github.com/11ty/eleventy-plugin-vue)/[React](https://github.com/signalkuppe/eleventy-react) |
| Assets directory | public | public | static | static | [Passthrough File Copy](https://www.11ty.dev/docs/copy/) |
| Distribution directory | dist | .next | public | public | _site |
| Plugin ecosystem | Yes | Yes | Yes | No | Yes |
| Routing | [pages](https://alephjs.org/docs/basic-features/pages)/[APIs](https://alephjs.org/docs/basic-features/apis) | pages | pages | pages/routes | Imperative API  |
| Image optimization | ? | [next/image](https://nextjs.org/docs/api-reference/next/image) | [gatsby-image](https://www.gatsbyjs.com/plugins/gatsby-image/)  [gatsby-plugin-sharp](https://www.gatsbyjs.com/plugins/gatsby-plugin-sharp/) | ? | [eleventy-img](https://github.com/11ty/eleventy-img) |
| Data Sourcing | useDeno/[APIs](https://alephjs.org/docs/basic-features/apis) | [getServerSideProps/ getStaticProps](https://nextjs.org/docs/basic-features/data-fetching) | [GraphQL](https://www.gatsbyjs.com/docs/content-and-data/) | custom express like server | [Data API](https://www.11ty.dev/docs/data/) |
| Serverless functions ready | Yes | Yes | Through plugins | ? | ? |
| Modules | ES Modules | CommonJS/ES Modules | CommonJS/ES Modules | CommonJS | CommonJS |
| Compiler | SWC | Babel | Babel | Babel | None |
| Bundle | ? | Webpack | Webpack | Webpack/Rollup | None |
| Service Worker | ? | [next-pwa](https://www.npmjs.com/package/next-pwa) | [gatsby-plugin-offline](https://www.gatsbyjs.com/plugins/gatsby-plugin-offline/) | [built-in](https://sapper.svelte.dev/docs#src_service-worker_js) | [eleventy-plugin-pwa](https://www.npmjs.com/package/eleventy-plugin-pwa) |
| Incremental builds | ? | server-side caching | Exclusive to [Gatsby Cloud](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/) | No | [Planned](https://github.com/11ty/eleventy/issues/108) |
| Preload links | Anchor / a tag | Link component | Link component | rel=prefetch attribute to normal HTML a tag | ? |
| PWA | Planned | [next-pwa](https://www.npmjs.com/package/next-pwa) | [gatsby-plugin-manifest](https://www.gatsbyjs.com/plugins/gatsby-plugin-manifest/) | ? | [eleventy-plugin-pwa](https://www.npmjs.com/package/eleventy-plugin-pwa) |
| AMP | Planned | [next/amp](https://nextjs.org/docs/api-reference/next/amp) | [gatsby-plugin-amp](https://www.gatsbyjs.com/plugins/gatsby-plugin-amp/)| ? | [@ampproject/eleventy-plugin-amp](https://www.npmjs.com/package/@ampproject/eleventy-plugin-amp)|

A comparison of SSG/SSR frameworks, it can be useful to compare all features in one place.
Add any feature you think it is not added here, when the table is complete it can be used as a markdown file for reference on how to migrate from other frameworks to Aleph.js/Deno.

**Comment by aadamsx** [comment]
You should have used SvelteKit instead of Sapper.
